src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultStates = [];
+
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      newState = { ...newState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
+    resultStates.push(newState);
+    currentState = newState;
```

Comment:
[DONT REPEAT YOURSELF]: You're creating a new `newState` object in each iteration of the loop. Instead, create it once outside the loop and then modify this copy within the `switch` cases. This will avoid unnecessary object creation and improve performance.